rfcs/rdm-0064-custom-fields.md

```diff
@@ -0,0 +1,636 @@
+- Applying permissions per-field.
+- Configuring custom fields using the UI, for example the _Backoffice_.
+- Compatibility with OpenSearch.
+- Add custom fields to existing sections, for example _Basic Information_.
+
+## Detailed design
+
+The topics covered by this RFC are the following:
+
+- [Defining custom fields for an instance](#defining-custom-fields-for-an-instance)
+- [Building new custom field types](#building-new-custom-field-types)
+- [Storing and indexing values for custom fields](#storing-and-indexing-values-for-custom-fields)
+- [Linking custom fields to vocabularies](#linking-custom-fields-to-vocabularies)
+- [Rendering widgets for custom fields in a form](#rendering-widgets-for-custom-fields-in-a-form)
+- [Building new custom widgets](#building-new-custom-widgets)
+- [Rendering custom fields values in the record landing page](#rendering-custom-fields-values-in-the-record-landing-page)
+- [Adding a custom field as facet in the search page](#adding-a-custom-field-as-facet-in-the-search-page)
+
+Along the RFC the following example use case will be used to exaplain the corresponding topic:
+
+_At CERN, I want to choose the experiment to which the record belong to, assuming experiments come from a controlled vocabulary. In addition, I want to be able to filter the record search by experiment._
+
+> Note: Along this RFC configuration variables are prefixed with `RDM_`, this means that custom fields are being added to _records_. However, the same configuration variables will be available in communities and will be used then adding custom fields to them. Namely, `COMMUNITIES_CUSTOM_FIELDS` and `COMMUNITIES_CUSTOM_FIELDS_UI`.
+
+### Defining custom fields for an instance
+
+The first step to add custom fields is to define/configure them. The instance administrator will change the configuraiton variables `RDM_CUSTOM_FIELDS` and `RDM_CUSTOM_FIELDS_UI` in the `invenio.cfg` configuration file. The following example shows only the configuration needed to **enable the custom field** in the data model not how to display them. This second part will be explained in the [Rendering widgets for custom fields in a form](#rendering-widgets-for-custom-fields-in-a-form) section.
+
+```python
+NAMESPACES = {
+    "cern": "https://cern.ch/terms",
+    "dwc": "http://rs.tdwg.org/dwc/terms/"
+}
+
+RDM_CUSTOM_FIELDS = {
+    "experiment": VocabularyCF(
+        vocabulary_id="cernexperiments",  # see namespacing in the unresolved questions section
+        name="experiment",
+        namespace="cern", # or other, for example, dwc for Darwin Core
+        required=True,
+        validation=custom_validate_func(),
+        error_messages={"required": "custom message"},
+    ),
+}
+```
+
+The last three attributes, namely `required`, `validation_func` and `error_messages`, are passed to the Marshmallow field. This allows the instance administrator to configure custom behaviour, both in validation and error messages returned.
+
+**Namespacing**
```

Comment:
This section could use some love. But I'm not so familiar with Zenodo's implementation. Anything you see missing @slint?

Reply:
I think the namespacing also affects the `RDM_CUSTOM_FIELDS` key uniqueness. E.g. if we had `cern:collectionID` and `dwc:collectionID`, what would be the keys in `RDM_CUSTOM_FIELDS`? Possible solutions:

- the keys would be `cern_collectionID`,  `dwc_collectionID` (not sure about the casing here, but again it's up to each instance probably)